sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/PerPartitionSdf.java

```diff
@@ -0,0 +1,152 @@
+      this.underlying = underlying;
+      this.lastClaimed = Optional.empty();
+    }
+
+    @Override
+    public boolean tryClaim(OffsetByteProgress position) {
+      boolean claimed = underlying.tryClaim(position);
+      if (claimed) {
+        lastClaimed = Optional.of(position.lastOffset());
+      }
+      return claimed;
+    }
+
+    @Override
+    public OffsetRange currentRestriction() {
+      return underlying.currentRestriction();
+    }
+
+    @Override
+    public @Nullable SplitResult<OffsetRange> trySplit(double fractionOfRemainder) {
+      return underlying.trySplit(fractionOfRemainder);
+    }
+
+    @Override
+    public void checkDone() throws IllegalStateException {
+      underlying.checkDone();
+    }
+
+    @Override
+    public IsBounded isBounded() {
+      return underlying.isBounded();
+    }
+  }
+
+  @GetInitialWatermarkEstimatorState
+  Instant getInitialWatermarkState() {
+    return Instant.EPOCH;
+  }
+
+  @NewWatermarkEstimator
+  MonotonicallyIncreasing newWatermarkEstimator(@WatermarkEstimatorState Instant state) {
+    return new MonotonicallyIncreasing(state);
+  }
+
+  @ProcessElement
+  public ProcessContinuation processElement(
+      RestrictionTracker<OffsetRange, OffsetByteProgress> tracker,
+      @Element SubscriptionPartition subscriptionPartition,
+      OutputReceiver<SequencedMessage> receiver,
+      BundleFinalizer finalizer)
+      throws Exception {
+    WrappedTracker wrapped = new WrappedTracker(tracker);
+    try (PartitionProcessor processor =
+        processorFactory.newProcessor(subscriptionPartition, wrapped, receiver)) {
+      processor.start();
+      ProcessContinuation result = processor.waitForCompletion(maxSleepTime);
+      wrapped.lastClaimed.ifPresent(
+          lastClaimedOffset ->
+              finalizer.afterBundleCommit(
+                  Instant.ofEpochMilli(Long.MAX_VALUE),
+                  () -> {
+                    Committer committer = committerFactory.apply(subscriptionPartition);
+                    committer.startAsync().awaitRunning();
+                    // Commit the next-to-deliver offset.
+                    committer.commitOffset(Offset.of(lastClaimedOffset.value() + 1)).get();
+                    committer.stopAsync().awaitTerminated();
+                  }));
+      return result;
+    }
+  }
+
+  @GetInitialRestriction
+  public OffsetRange getInitialRestriction(@Element SubscriptionPartition subscriptionPartition) {
+    try (InitialOffsetReader reader = offsetReaderFactory.apply(subscriptionPartition)) {
+      Offset offset = reader.read();
+      return new OffsetRange(offset.value(), Long.MAX_VALUE /* open interval */);
+    }
+  }
+
+  @NewTracker
+  public RestrictionTracker<OffsetRange, OffsetByteProgress> newTracker(
+      @Element SubscriptionPartition subscriptionPartition, @Restriction OffsetRange range) {
+    return trackerFactory.apply(subscriptionPartition, range);
```

Comment:
I'm curious why we need a `trackerFactory` here instead of returning your `OffsetByteRangeTracker` directly. Do you expect your SDF user to implement their own restriction tracker?